### PR TITLE
Rely on `$request->input()` instead of manually json decoding request body

### PR DESCRIPTION
**Pull request by pyrou**
- [x] Added automated tests (All existing unit tests cover well each cases)
- [x] Documented for all relevant versions
- [x] Updated the changelog

**Changes**

With this changes, on Laravel projects using Http middleware that manipulate the request payload (such as `Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull`, `Illuminate\Foundation\Http\Middleware\TrimStrings` or `Illuminate\Foundation\Http\Middleware\TransformsRequest`) will now work as expected. This is the primary goal of this pull request.

**Breaking changes**

Having middleware in place than suddenly taking effect, can introduce breaking changes on your project. Make sure you adjust your middleware & middlewareGroups in your `app/http/kernel.php` according to your needs.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 /.phpunit.result.cache
 /.php-cs-fixer.cache
 /composer.lock
+.idea
```
Comment (spawnia) [inline .gitignore:6]:
```suggestion
/.idea
```

Keep it consistent

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+- Rely on $request->input() instead of manually json decoding request body https://github.com/laragraph/utils/pull/12
```
Comment (spawnia) [inline CHANGELOG.md:12]:
```suggestion
- Rely on `$request->input()` instead of manually json decoding request body https://github.com/laragraph/utils/pull/12
```

File: src/RequestParser.php [hunk 3]
```diff
@@ -118,4 +85,38 @@ protected function inlineFiles(Request $request): array
...
+         */
+        $contentType = $request->header('Content-Type');
+
+        if (Str::startsWith($contentType, 'multipart/form-data')) {
+            return $this->inlineFiles($request);
+        }
+
+        $bodyParams = $request->input();
+
+        if (is_array($bodyParams) && Arr::isAssoc($bodyParams)) {
```
Comment (spawnia) [inline src/RequestParser.php:109]: Should we throw if this is a list?

File: src/RequestParser.php [hunk 4]
```diff
@@ -118,4 +85,38 @@ protected function inlineFiles(Request $request): array
...
+        }
+
+        if (Str::startsWith($contentType, 'application/graphql')) {
+            /** @var string $content */
+            $content = $request->getContent();
+
+            return ['query' => $content];
+        }
+
+        throw new BadRequestGraphQLException('Could not parse GraphQL request body');
```
Comment (spawnia) [inline src/RequestParser.php:120]: Can we give the user a better indication as to why we could not parse it?

**Review by spawnia — CHANGES_REQUESTED** [review]

**Comment by pyrou** [comment]
@mfn Hi, yes, as initialy proposed by @spawnia, the idea was to propose a new major version that Lighthouse 6 would support alongside current Laragraph version. New comers would benefit latest version when doing a first `composer require` while giving time to upgrade to others.

File: src/RequestParser.php (hunk 3 above)
Comment (pyrou) [reply src/RequestParser.php:109]: As shown by test `testNonArrayJson()`

Calling 
```php
$request = $this->makeRequest(
    'POST', [], [],
    ['Content-Type' => 'application/json'],
    '"this should be a map with query, variables, etc."'
);
```

Will result in a $bodyParams containing :
```php
array:1 [
  0 => "this should be a map with query, variables, etc."
]
```

File: src/RequestParser.php (hunk 4 above)
Comment (pyrou) [reply src/RequestParser.php:120]: Since we don't do anymore the json_decode ourselves, this became probably out of scope. This could only occur on malformed payload (eg. sending json data with a application/x-www-form-urlencoded header).
I'm not sure on how to improve this message here. Would be happy if you have any suggestion .

Comment (spawnia) [reply src/RequestParser.php:120]: We might include the content-type header at least.

File: src/RequestParser.php (hunk 3 above)
Comment (spawnia) [reply src/RequestParser.php:109]: I think in that case we should definitely throw! Previously, this would have resulted in `'GraphQL Server expects JSON object or array, but got ...`, much better than the generic error message that is now thrown. We should probably extend the tests to ensure that sensical error messages are produced in those edge cases.

Comment (pyrou) [reply src/RequestParser.php:110]: I added two explicit exception to cover previous scope. Honestly not a big fan at all to provide back to client the given request payload, somehow this could lead to security issue, xss injection etc..

File: tests/Unit/RequestParserTest.php [hunk 5]
```diff
@@ -148,7 +149,6 @@ public function nonsensicalContentTypes(): iterable
         yield ['application/foobar'];
         yield ['application/josn'];
         yield ['application/grapql'];
-        yield ['application/foo;charset=application/json'];
```
Comment (pyrou) [inline tests/Unit/RequestParserTest.php:151]: In laravel, $request->isJson() for the following header returns true, because only /json or +json substring is checked in the Content-Type value

See https://github.com/laravel/framework/blob/9.x/src/Illuminate/Http/Concerns/InteractsWithContentTypes.php#L14-L17

File: src/RequestParser.php [hunk 6]
```diff
@@ -27,53 +30,18 @@ public function __construct()
     /**
      * Converts an incoming HTTP request to one or more OperationParams.
      *
-     * @throws \Laragraph\Utils\BadRequestGraphQLException
+     * @throws RequestError
+     * @throws BadRequestGraphQLException
      *
-     * @return \GraphQL\Server\OperationParams|array<int, \GraphQL\Server\OperationParams>
+     * @return OperationParams|array<int, OperationParams>
```
Comment (spawnia) [inline src/RequestParser.php:36]: Always use fqcn here

File: tests/Unit/RequestParserTest.php (hunk 5 above)
Comment (spawnia) [reply tests/Unit/RequestParserTest.php:151]: Stuff like this is why I preferred doing this ourselves...

File: src/RequestParser.php [hunk 7]
```diff
@@ -27,53 +28,18 @@ public function __construct()
     /**
      * Converts an incoming HTTP request to one or more OperationParams.
      *
-     * @throws \Laragraph\Utils\BadRequestGraphQLException
+     * @throws \GraphQL\Server\RequestError
+     * @throws \GraphQL\Server\BadRequestGraphQLException
```
Comment (spawnia) [inline src/RequestParser.php:32]:
```suggestion
     * @throws \Laragraph\Utils\BadRequestGraphQLException
```

**Comment by spawnia** [comment]
Thanks, released with https://github.com/laragraph/utils/releases/tag/v2.0.0
